include/deal.II/lac/dynamic_sparsity_pattern.h

```diff
@@ -1040,14 +1040,25 @@ DynamicSparsityPattern::begin (const size_type r) const
   // pattern
   //
   // note: row_length(row) returns zero if the row is not locally stored
-  //
-  // TODO: this is way too slow when used in parallel, so do not use it on
-  // non-owned rows
   size_type row = r;
-  while ((row<n_rows())
-         &&
-         (row_length(row)==0))
-    ++row;
+
+  size_type first_own_row = (rowset.size() > 0) ?
+                            rowset.nth_index_in_set(0)
+                            : 0;
+  size_type last_own_row = (rowset.size() > 0) ?
+                           rowset.nth_index_in_set(rowset.n_elements()-1)
+                           : n_rows();
+
+  // skip to first locally owned row
+  if (row<first_own_row)
+    row = first_own_row;
+
+  while (row<n_rows() && row_length(row)==0)
+    {
+      ++row;
+      if (row > last_own_row)
+        row = n_rows();
+    }
```

Comment:
Do you think that this is enough? It may catch simple cases, but I fear that it is not a general solution. I think of a mesh of a uniformly refined square with the number of processors divisble by four. The access over the middle vertex will jump very far in terms of index because it jumps over more than half of the processors and half of the index space at some point in between. I think we need to augment things a bit here: If row_length is zero, we would need to do something like a `lower_bound` on the index set.